### PR TITLE
[percep] Added config file

### PR DESCRIPTION
**Pull request by jubileew**
Created config file, added variables to artag_detector.cpp/hpp and pcl.cpp/hpp

## Compile Test
- [x] Offline Debug: ` ./jarvis build jetson/percep -o with_zed=false ar_detection=true obs_detection=true perception_debug=true `
  - [x] No compiler errors or warnings
- [x] Offline Silent: ` ./jarvis build jetson/percep -o with_zed=false ar_detection=true obs_detection=true perception_debug=false `
  - [x] No compiler errors or warnings
- [x] ZED Debug: ` ./jarvis build jetson/percep -o with_zed=true ar_detection=true obs_detection=true perception_debug=true `
  - [x] No compiler errors, only ZED warnings
- [x] ZED Silent: ` ./jarvis build jetson/percep -o with_zed=true ar_detection=true obs_detection=true perception_debug=false `
   - [x] No compiler errors, only ZED warnings

## Offline Debug Test
- [x] `./jarvis build jetson/percep -o with_zed=false ar_detection=true obs_detection=true perception_debug=true`
- [x] Run code against test1 image set located in perception-files repository
- Check for:
   - [x]  Left and right path are being calculated
   - [x]  Calculated bearing is reasonable given viewer output
   - [x]  AR Tag ID is 2
   - [x]  Obstacles with large width are detected
   - [x]  Path is projected around not through large obstacles
   - [x]  Short obstacles are detected
   - [x]  Steep inclines are detected as obstacles
- [x] Run code against test2 image set located in perception-files repository
- Check for:
   - [x]  Left and right path are being calculated
   - [x]  Calculated bearing is reasonable given viewer output
   - [x]  AR Tag ID is 6
   - [x]  2 AR Tags detected simultaneously
   - [x]  Obstacles with large width are detected
   - [x]  Path is projected around not through large obstacles
   - [x]  Short obstacles are detected
   - [x]  Steep inclines are detected as obstacles

## Offline Silent Test
- [x] `./jarvis build jetson/percep -o with_zed=false ar_detection=true obs_detection=true perception_debug=false`
- [x] Run `time ./jarvis exec jetson/percep` against test_case1 image set
- [x] Run `time ./jarvis exec jetson/percep` against test_case2 image set
- [ ] User value is less than or equal to \<TODO Determine Value\>

## Unit Test
- [x] ` ./jarvis build jetson/percep -o with_zed=true ar_detection=true obs_detection=true perception_debug=true `
- [x] `./jarvis exec lcm_tools_echo Obstacle "/obstacle"`
- [x] `./jarvis exec lcm_tools_echo TargetList "/target_list"`
- Test environment contains:
  - [x] 1 large obstacle ~ over 40cm tall
  - [x] 2 small obstacles ~  25cm-30cm tall
  - [x] 1 post
  - [x] 1 hill
- Test for:
  - [x] Detected large obstacle 6m away
  - [x] Detected small obstacle 2m away
  - [x] Detected clear path when 1.2m gap between 1 large and 1 small obstacle
  - [x] Left and right path are being calculated
  - [x] Calculated bearing is reasonable given viewer output of path
  - [x] AR Tag ID is the correct ID
  - [x] Obstacles with large width are detected
  - [x] Short obstacles are detected
  - [ ] FPS values don't exceed \<TODO find this value in Issue #552\>
  - [x] Obstacle.lcm in /obstacle is populated correctly
  - [x] TargetList.lcm in /target_list is populated correctly

## Integration Test
- [x] Create new branch based off of main
- [x] Checkout new branch
- [x] Cherry-Pick all PRs that need testing onto new branch
- [x] Compile all Software Packages
- [x] Run all code simultaneously
- [x] Obstacle.lcm in /obstacle is populated correctly
- [x] TargetList.lcm in /target_list is populated correctly
- [x] Sensors are Active
  - [x] Stereo Camera (ZED)
  - [x] Inertial Measurement Unit (IMU)
  - [x] Global Positioning System (GPS)

### CODE REVIEW

File: jetson/percep/deprecated/calibrate/main.cpp [hunk 1]
```diff
@@ -72,7 +72,7 @@ int main(){
 	    Mat depth_img;
 
 		const float roverWidth = 46 * 25.4; //inches to mm
-		const float focalLength = 2.8; //zed focal length in mm
+		const float focalLength = mRoverConfig["zed_specs"]["focalLength"].GetFloat(); //zed focal length in mm
```
Comment (mmartin4972) [inline jetson/percep/deprecated/calibrate/main.cpp:75]: If you change this back I think you will be able to pass Travis.

**Review by mmartin4972 — CHANGES_REQUESTED** [review]

File: jetson/percep/artag_detector.cpp [hunk 2]
```diff
@@ -35,9 +40,9 @@ TagDetector::TagDetector() {  //initializes detector object with pre-generated d
 
     // initialize other special parameters that we need to properly detect the URC (Alvar) tags
     alvarParams = new cv::aruco::DetectorParameters();
-    alvarParams->markerBorderBits = 2;
-    alvarParams->doCornerRefinement = false;
-    alvarParams->polygonalApproxAccuracyRate = 0.08;
+    alvarParams->markerBorderBits = MARKER_BORDER_BITS;
+    alvarParams->doCornerRefinement = DO_CORNER_REFINEMENT;
+    alvarParams->polygonalApproxAccuracyRate =POLYGONAL_APPROX_ACCURACY_RATE;
```
Comment (mmartin4972) [inline jetson/percep/artag_detector.cpp:45]: nit: add space before P

File: jetson/percep/artag_detector.cpp [hunk 3]
```diff
@@ -16,8 +15,14 @@ void onMouse(int event, int x, int y, int flags, void *userdata) {
...
-
+TagDetector::TagDetector(const rapidjson::Document &mRoverConfig) //initializes detector object with pre-generated dictionary of tags
+    :  //Populate Constants from Config File
+   BUFFER_ITERATIONS{mRoverConfig["buffer_iterations"].GetInt()},
+   MARKER_BORDER_BITS{mRoverConfig["alvar_params"]["markerBorderBits"].GetInt()},
+   DO_CORNER_REFINEMENT{!!mRoverConfig["alvar_params"]["doCornerRefinement"].GetInt()},
+   POLYGONAL_APPROX_ACCURACY_RATE{mRoverConfig["alvar_params"]["polygonalApproxAccuracyRate"].GetDouble()},
+   MM_PER_M{mRoverConfig["mm_per_m"].GetInt()},
+   DEFAULT_TAG_VAL{mRoverConfig["DEFAULT_TAG_VAL"].GetInt()} {
     cv::FileStorage fsr("jetson/percep/alvar_dict.yml", cv::FileStorage::READ);
```
Comment (mmartin4972) [inline jetson/percep/artag_detector.cpp:26]: Can we format this better? Maybe move the colon up a line, make the comment on line 19 in line with all of the variable declarations, and add a new line after the open curly bracket on line 25?

File: jetson/percep/main.cpp [hunk 4]
```diff
@@ -8,10 +8,27 @@ using namespace cv;
...
+   string config = "";
+   string setting;
+   while( configFile >> setting )
+       {
+          config += setting;
+       }
+   configFile.close();
+ 
+   mRoverConfig.Parse( config.c_str() );
+
```
Comment (mmartin4972) [inline jetson/percep/main.cpp:28]: Can you fix the indentation here? Also, can you change the comment on the top to be formatted the same as the other sections are?

File: jetson/percep/artag_detector.cpp (hunk 3 above)
Comment (mmartin4972) [reply jetson/percep/artag_detector.cpp:26]: Checkout how we did the formatting of the constructor in PCL.cpp. That formatting looks better

File: jetson/percep/pcl.cpp [hunk 5]
```diff
@@ -3,7 +3,30 @@
...
+        pt_cloud_ptr{new pcl::PointCloud<pcl::PointXYZRGB>} {
+
+        #if ZED_SDK_PRESENT
+        sl::Resolution cloud_res = sl::Resolution(PT_CLOUD_WIDTH, PT_CLOUD_HEIGHT);
+        cloudArea = cloud_res.area();
+        #else
+        cloudArea = PT_CLOUD_WIDTH*PT_CLOUD_HEIGHT;
+        #endif
+
+    };
```
Comment (mmartin4972) [inline jetson/percep/pcl.cpp:29]: Can you fix the indentation here?

Comment (mmartin4972) [reply jetson/percep/pcl.cpp:29]: I think either everything needs to be shifted left or the comment needs to be shifted right

**Review by mmartin4972 — CHANGES_REQUESTED** [review]

File: jetson/percep/pcl.cpp [hunk 6]
```diff
@@ -3,7 +3,30 @@
...
+        UP_BD_Z{mRoverConfig["pt_cloud"]["pass_through"]["upperBdZ"].GetDouble()},
+        UP_BD_Y{mRoverConfig["pt_cloud"]["pass_through"]["upperBdY"].GetDouble()},
+        LOW_BD{mRoverConfig["pt_cloud"]["pass_through"]["lowerBd"].GetDouble()},
+
+        //Other Values
+        bearing{0}, distance{0}, detected{false},
+        pt_cloud_ptr{new pcl::PointCloud<pcl::PointXYZRGB>} {
+
+        #if ZED_SDK_PRESENT
+        sl::Resolution cloud_res = sl::Resolution(PT_CLOUD_WIDTH, PT_CLOUD_HEIGHT);
```
Comment (mmartin4972) [inline jetson/percep/pcl.cpp:23]: Anything within the #if statements should be indented. Could you do that here?

**Review by mmartin4972 — CHANGES_REQUESTED** [review]

File: jetson/percep/pcl.cpp [hunk 7]
```diff
@@ -3,7 +3,30 @@
 
 #if OBSTACLE_DETECTION
 
-const int MAX_FIELD_OF_VIEW_ANGLE = 70;
+ //Constructor
```
Comment (mmartin4972) [inline jetson/percep/pcl.cpp:6]: nit: Can you align this comment with the constructor?

File: jetson/percep/pcl.hpp [hunk 8]
```diff
@@ -49,20 +49,32 @@ class compareLine {
...
+     * RANSAC SECTION in pt_cloud section:
+     *      max iterations
+     *      segmentation epsilon
+     *      distance threshold
+     * 
+     * Euclidean cluster section in pt_cloud section
+     *      cluster tolerance
+     *      min cluster size
+     *      max cluster size
+     * */
```
Comment (mmartin4972) [inline jetson/percep/pcl.hpp:75]: Can you delete the TODO?

File: jetson/percep/artag_detector.cpp [hunk 9]
```diff
@@ -148,24 +157,23 @@ void TagDetector::updateDetectedTagInfo(rover_msgs::Target *arTags, pair<Tag, Ta
     tags.buffer.push_back(0);
     tags.buffer.push_back(0);
 
-  for (uint i=0; i<2; i++){
-    if(tags.id[i] == -1){//no tag found
-        if(tags.buffer[i] <= 20){//send buffered tag until tag is found
+  for (uint i=0; i<2; i++)
+  {
```
Comment (mmartin4972) [inline jetson/percep/artag_detector.cpp:161]: nit: Could you fix the placement of this curly brace to abide by our style guide?

**Review by mmartin4972 — CHANGES_REQUESTED** [review]

File: config/percep/project.ini [hunk 10]
```diff
@@ -0,0 +1,2 @@
+[build]
+lang=config
```
Comment (mmartin4972) [inline config/percep/project.ini:2]: Could you add a line to the end of this file?

File: jetson/percep/pcl.hpp [hunk 11]
```diff
@@ -50,20 +50,31 @@ class compareLine {
...
+    double ROVER_W_MM;
+    float LEAF_SIZE;
+    //RANSAC constants
+    int MAX_ITERATIONS;
+    double SEGMENTATION_EPSLION;
+    double DISTANCE_THRESHOLD;
+    //Euclidean cluster constants
+    int CLUSTER_TOLERANCE;
+    int MIN_CLUSTER_SIZE;
+    int MAX_CLUSTER_SIZE;
```
Comment (mmartin4972) [inline jetson/percep/pcl.hpp:71]: There seemed to be some compiler warnings about the order that you initialize these values. Could you get rid of those warnings? Should just have to do with reordering the variable definitions.

**Review by mmartin4972 — CHANGES_REQUESTED** [review]

**Review by DavidSmith166 — APPROVED** [review]

**Review by raytitan — COMMENTED** [review]
Please remove the changes unrelated to perception - the Dockerfile and rapidjson

File: Dockerfile [hunk 12]
```diff
@@ -141,6 +141,7 @@ RUN echo "!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!INSTALLING EIGEN" && \
     make install && \
     cd ../.. && sudo rm -rf eigen-3.3.7/ && sudo rm -f eigen-3.3.7.tar.gz
 
+<<<<<<< HEAD
```
Comment (mmartin4972) [inline Dockerfile:144]: Delete this line

File: Dockerfile [hunk 13]
```diff
@@ -154,3 +155,6 @@ RUN echo "!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!INSTALLING PCL" && \
     make -j1 && make install && ldconfig && \
     cd /usr/local && \
     rm -rf pcl-pcl-1.11.1
+=======
```
Comment (mmartin4972) [inline Dockerfile:158]: Delete 158-160

File: jetson/percep/camera.cpp [hunk 14]
```diff
@@ -327,9 +327,9 @@ cv::Mat Camera::Impl::depth() {
 }
 
 void Camera::record_ar_init() {
-    //initializing ar tag videostream object
```
Comment (mmartin4972) [inline jetson/percep/camera.cpp:330]: Could you add this comment back in?

File: jetson/percep/camera.hpp [hunk 15]
```diff
@@ -1,6 +1,8 @@
 #pragma once
 #include <opencv2/opencv.hpp>
 #include "perception.hpp"
+#include "rapidjson/document.h"
```
Comment (mmartin4972) [inline jetson/percep/camera.hpp:4]: Could you include these in perception.hpp? Should keep you from having to repeatedly include them in other files

File: jetson/percep/pcl.hpp [hunk 16]
```diff
@@ -79,7 +81,8 @@ class PCL {
         viewer_original -> close();
       #endif
     };
-
+    
+    //Member Variables
     double bearing;
```
Comment (mmartin4972) [inline jetson/percep/pcl.hpp:86]: Could you move these up with the other variables?

File: jetson/percep/pcl.hpp [hunk 17]
```diff
@@ -52,25 +52,27 @@ class PCL {
         shared_ptr<pcl::visualization::PCLVisualizer> viewer;
```
Comment (mmartin4972) [inline jetson/percep/pcl.hpp:52]: The indentation of these two lines doesn't look consistent with the rest of the file. Could you fix that?

**Review by mmartin4972 — CHANGES_REQUESTED** [review]

File: jetson/percep/camera.hpp (hunk 15 above)
Comment (mmartin4972) [reply jetson/percep/camera.hpp:3]: Did you try this and it not work? It looks like you deleted it and then added it back in. Could you add it to Perception.hpp?

File: jetson/percep/camera.cpp [hunk 18]
```diff
@@ -380,8 +381,7 @@ void Camera::Impl::dataCloud(pcl::PointCloud<pcl::PointXYZRGB>::Ptr &p_pcl_point
 
 #endif
 
-Camera::Camera(const rapidjson::Document &config) : impl_{new Camera::Impl}, rgb_foldername{""},
-                   depth_foldername{""}, pcl_foldername{""} , mRoverConfig( config ) {}
+Camera::Camera(const rapidjson::Document &config) : impl_{new Camera::Impl}, rgb_foldername{""}, depth_foldername{""}, pcl_foldername{""} , mRoverConfig( config ) {}
```
Comment (mmartin4972) [inline jetson/percep/camera.cpp:384]: Could you break this line into two lines?

File: jetson/percep/pcl.hpp [hunk 19]
```diff
@@ -52,42 +52,42 @@ class PCL {
...
-    float LEAF_SIZE;
-    //RANSAC constants
-    int MAX_ITERATIONS;
-    double SEGMENTATION_EPSLION;
-    double DISTANCE_THRESHOLD;
-    //Euclidean cluster constants
-    int CLUSTER_TOLERANCE;
-    int MIN_CLUSTER_SIZE;
-    int MAX_CLUSTER_SIZE;
+        //Constants
```
Comment (mmartin4972) [inline jetson/percep/pcl.hpp:55]: nit: could you break these sections of variables up with newlines?

**Review by mmartin4972 — DISMISSED** [review]

**Review by vidyasilai — APPROVED** [review]

File: jetson/percep/camera.hpp (hunk 15 above)
Comment (jubileew) [reply jetson/percep/camera.hpp:3]: I have it in both files, but for some reason when I remove it from camera.cpp it doesn't compile

Comment (mmartin4972) [reply jetson/percep/camera.hpp:3]: That's strange. We can just leave it for now

**Review by mmartin4972 — APPROVED** [review]

**Review by vidyasilai — APPROVED** [review]

File: config/percep/config.json [hunk 20]
```diff
@@ -0,0 +1,58 @@
...
+    },
+
+    "alvar_params":
+    {
+        "markerBorderBits": 2,
+        "doCornerRefinement": 0,
+        "polygonalApproxAccuracyRate": 0.08
+
+    }
+}
```
Comment (vidyasilai) [inline config/percep/config.json:58]: More of a question for me than it is a comment for you, but is there a reason why some constants use snake case and a few use camel case? Thought we were moving towards camel case in terms of style guidelines but is this just for the config file in particular?

File: config/percep/config.json [hunk 21]
```diff
@@ -0,0 +1,58 @@
+{
+    "frame_write_interval": 10,
+    "buffer_iterations": 20,
+    "DEFAULT_TAG_VAL": -1,
```
Comment (DavidSmith166) [inline config/percep/config.json:4]: NIT: why is this all caps when the other keys are not?

File: config/percep/config.json [hunk 22]
```diff
@@ -0,0 +1,58 @@
...
+        "downsample_voxel_filter": 20.0,
+       
+        "ransac": {
+            "max_iterations": 400,
+            "segmentation_epsilon": 10,
+            "distance_threshold": 100
+        },
+
+        "pass_through": {
+            "upperBdZ": 7000.0,
```
Comment (DavidSmith166) [inline config/percep/config.json:25]: NIT: what does the switch from snake to camel case signify? If there is no difference then use only snake or only camel case

File: config/percep/config.json [hunk 23]
```diff
@@ -0,0 +1,58 @@
+{
+    "frame_write_interval": 10,
+    "buffer_iterations": 20,
+    "DEFAULT_TAG_VAL": -1,
+    "mm_per_m":1000,
+    "threshold_confidence":90,
```
Comment (DavidSmith166) [inline config/percep/config.json:6]: NIT: these items seem to have a specific use but are not grouped together like other items, why is that?

**Review by DavidSmith166 — COMMENTED** [review]
Good work, I have some style and organization concerns. @mmartin4972 what are your thoughts?

**Review by DavidSmith166 — DISMISSED** [review]

File: config/percep/config.json (hunk 20 above)
Comment (mmartin4972) [reply config/percep/config.json:67]: Good point. We'll clean the style up in PR #576

**Review by mmartin4972 — APPROVED** [review]

**Review by vidyasilai — DISMISSED** [review]

**Review by mmartin4972 — APPROVED** [review]

**Review by vidyasilai — APPROVED** [review]

**Review by DavidSmith166 — APPROVED** [review]

**Pull request merged**